### ISSUE TITLE
Behavior difference in System.Uri.IsAbsoluteUri on some platforms when UriKind = UriKind.RelativeOrAbsolute

### ISSUE THREAD

**Issue by sepidehMS** [dotnet/corefx#20046]
The following code snippet fails on Ubuntu<s>14.04</s>, OSX, PortableLinux, and CentOS, while it passes on Windows <s>and Ubuntu16.04.</s>
```c#
string path = Path.Combine(Path.GetTempPath());
Uri uri = new Uri(path, UriKind.RelativeOrAbsolute);
Assert.True(uri.IsAbsoluteUri);
```
This was discovered in https://github.com/dotnet/corefx/pull/19735 when investigating a XmlSchema validation test failure. [In case it helps: XmlSchemaSet creates such Uri in [here](https://source.dot.net/#System.Private.Xml/System/Xml/Schema/XmlSchemaSet.cs,368) and later tries to resolve a file in that uri location in [XmlResolver](https://source.dot.net/#System.Private.Xml/System/Xml/XmlResolver.cs,36)]

**Comment by sepidehMS** [comment]
@danmosemsft Who would be the right person to look at this?

**Comment by danmosemsft** [comment]
@karelz who owns URI?

**Comment by karelz** [comment]
It's in Networking area (although some people disagree). I updated [areas](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas) to clarify it.

@stephentoub can you please help take a look?

**Comment by stephentoub** [comment]
> The following code snippet fails on Ubuntu14.04, OSX, PortableLinux, and CentOS, while it passes on Windows and Ubuntu16.04.

@sepidehMS, can you share the links to the CI legs where you observed the differences amongst different Unix platforms?

cc: @tmds

**Comment by tmds** [comment]
If the Uri class is passed a path that starts with a '/' (e.g. /tmp) and kind is RelativeOrAbsolute, then on Windows and Unix the resulting uri is a relative uri.

**Comment by stephentoub** [comment]
> If the Uri class is passed a path that starts with a '/' (e.g. /tmp) and kind is RelativeOrAbsolute, then on Windows and Unix the resulting uri is a relative uri.

@tmds, yup.  My two questions:
1. Why did @sepidehMS see a different answer on Ubuntu16.04?  That should only be the case if GetTempPath() returned something that didn't begin with a '/', which would be very strange.
2. Why does Mono return true for this?  I tested on 4.2; maybe it's changed since then?

**Comment by tmds** [comment]
perhaps an envvar is overriding the temp path on ubuntu, indeed strange if it would not start with a /...
for mono: http://www.mono-project.com/docs/faq/known-issues/urikind-relativeorabsolute/

**Comment by danmosemsft** [comment]
@stephentoub I kicked off https://github.com/dotnet/corefx/pull/19735 again which should give us logs.

**Comment by danmosemsft** [comment]
I threw up a little test to see what each OS does.

**Comment by danmosemsft** [comment]
Note that `Path.Combine` passed a single parameter just returns that parameter unchanged.

All Linux distros and OSX fail. (You mentioned Ubuntu 16.04 passed but perhaps you were led astray becaues the default version of this job is ARM and does not run tests. I ran a regular job and it [failed](https://ci.dot.net/job/dotnet_corefx/job/master/job/ubuntu16.04_release_prtest/2/testReport/junit/System.PrivateUri.Tests/UriTests/TestCtor_String2/))

On non-Windows, temp path is `/tmp/` and on Windows it's `c:\users\dotnet-bot\appdata\local\temp`. No surprise. And
```
new Uri('/tmp/', UriKind.RelativeOrAbsolute).IsAbsoluteUri; // returns false
new Uri('c:\users\dotnet-bot\appdata\local\temp', UriKind.RelativeOrAbsolute).IsAbsoluteUri; // returns true
```
again presumably correct behavior.

@sepidehMS I'm going to close this issue as URI is behaving correctly, @ViktorHofer please open an issue to track whether `XmlSchemaSet` is expected to behave differently for relative vs. absolute URI as observed in https://github.com/dotnet/corefx/pull/19735.

**Comment by krwq** [comment]
@danmosemsft I believe that `new Uri(@"/tmp/", UriKind.RelativeOrAbsolute).IsAbsoluteUri; // returns false` should return true - AFAIK anything starting with slash on linux is considered an absolute path. IMO this is Uri bug

**Comment by stephentoub** [comment]
> AFAIK anything starting with slash on linux is considered an absolute path

Based on what?

**Comment by karelz** [comment]
Triage: We don't think there is anything left for 2.0.

@krwq if you have further evidence, please reply here (without reopening).

**Comment by tmds** [comment]
@karelz @stephentoub @sepidehMS This issue was now 'fixed' by converting the unix path to an absolute uri in the tests (https://github.com/dotnet/corefx/commit/3a3ead6abb6c27364f12e41478a34d79ced6fdd1#diff-0bf8d1637c34db29730c199a32fd08ff). I think it makes sense to add this behavior to the XmlReader.Create methods (on Unix platforms for urls starting with '/'). What do you think?

**Comment by krwq** [comment]
@tmds - XML's parameter (baseURI) is taking URI - I do not know the URI's spec enough to tell what is the correct behavior but IMO if this issue is by design then XML should match this behavior by using the defaults. If it was taking a path then I could 100% agree

**Comment by tmds** [comment]
@krwq The XmlReader.Create methods accept the uri as a string. My suggestion is to make existing code (like the one in the test) to keep working without modifications. This is by converting the string uri passed to the XmlReader.Create on Unix platforms when it starts with a '/' to an absolute file:// uri.

**Comment by krwq** [comment]
@tmds, I think it's better to stay consistent and it is not a good idea to do one thing in case of URI and different in XML. My intuition tells me that this issue should be the one to fix and I'd rather have different behavior on linux vs Windows for the same string than having to special case for each component returning paths.

**Comment by stephentoub** [comment]
We still have this workaround in place:
https://github.com/dotnet/corefx/pull/12490
At the time it was added exactly to fix the XmlReader.Create case, and it did.  Did something change since then so that it no longer does?

**Comment by krwq** [comment]
cc: @sepidehMS

**Comment by sepidehMS** [comment]
@stephentoub This issue is happening in cases where the baseUri is not null or empty string and so, since Uri.IsAbsoluteUri(baseUri) returns false for a baseUri such as `/tmp/` on Unix, it goes [here](https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L56). I'm not completely familiar with all possible cases for URI, but if the uri.StartsWith("/") is going to work correctly for all cases, we still have to #ifdef the code for Windows and Unix, if want to fix this in Xml side.

**Comment by tmds** [comment]
@stephentoub @sepidehMS @krwq The XmlReader.BaseUri property is sometimes set as a result of calling XmlResolver.ResolveUri and sometimes it is a string value assignment. It's these latter cases that are causing the failure for unix paths. The code has some comments which say it is safe to to the ResolveUri in that particular case. This implies that it may be unsafe to do it everywhere. I don't understand when it would be unsafe so it makes me wary of touching the code.

**Comment by tmds** [comment]
```C#
using System;
using System.IO;
using System.Xml;

namespace xmlbaseuri
{
    class Program
    {
        private const string XmlFilePath = "/tmp/file.xml";

        private static XmlParserContext CreateParserContext()
        {
            return new XmlParserContext(null, null, null, null, null, null, XmlFilePath, null, XmlSpace.None);
        }

        static void Main(string[] args)
        {
            File.WriteAllText(XmlFilePath, @"<?xml version=""1.0"" encoding=""UTF-8""?>");
            var factories = new (string Name, Func<XmlReader> Create)[]
            {
                ("XmlReader_Url       ", () => XmlReader.Create(XmlFilePath)),
                ("XmlReader_UrlCtx    ", () => XmlReader.Create(XmlFilePath, null, CreateParserContext())),
                ("XmlReader_StreamUrl ", () => XmlReader.Create(File.OpenRead(XmlFilePath), null, XmlFilePath)),
                ("XmlReader_StreamCtx ", () => XmlReader.Create(File.OpenRead(XmlFilePath), null, CreateParserContext())),
                ("XmlReader_TxtUrl    ", () => XmlReader.Create(new StreamReader(File.OpenRead(XmlFilePath)), null, XmlFilePath)),
                ("XmlReader_TxtCtx    ", () => XmlReader.Create(new StreamReader(File.OpenRead(XmlFilePath)), null, CreateParserContext())),
                ("TxtReader_Url       ", () => new XmlTextReader(XmlFilePath)),
                ("TxtReader_StreamUrl ", () => new XmlTextReader(XmlFilePath, File.OpenRead(XmlFilePath))),
                ("TxtReader_StreamCtx ", () => new XmlTextReader(File.OpenRead(XmlFilePath), XmlNodeType.Document, CreateParserContext())),
                ("TxtReader_StringCtx ", () => new XmlTextReader(File.ReadAllText(XmlFilePath), XmlNodeType.Document, CreateParserContext())),
                ("TxtReader_TxtUrl    ", () => new XmlTextReader(XmlFilePath, new StreamReader(File.OpenRead(XmlFilePath)))),
            };
            foreach (var factory in factories)
            {
                try
                {
                    using (var reader = factory.Create())
                    {
                        Console.WriteLine($"{factory.Name}: {reader.BaseURI}");
                    }
                }
                catch (Exception e)
                {
                    System.Console.WriteLine($"{factory.Name}: throws {e.Message}");
                }
            }
        }
    }
}
```
outputs:
```
XmlReader_Url       : file:///tmp/file.xml
XmlReader_UrlCtx    : throws BaseUri must be specified either as an argument of XmlReader.Create or on the XmlParserContext. If it is specified on both, it must be the same base URI. Line 1, position 1.
XmlReader_StreamUrl : /tmp/file.xml
XmlReader_StreamCtx : /tmp/file.xml
XmlReader_TxtUrl    : /tmp/file.xml
XmlReader_TxtCtx    : /tmp/file.xml
TxtReader_Url       : file:///tmp/file.xml
TxtReader_StreamUrl : /tmp/file.xml
TxtReader_StreamCtx : /tmp/file.xml
TxtReader_StringCtx : /tmp/file.xml
TxtReader_TxtUrl    : /tmp/file.xml
```

**Comment by danmosemsft** [comment]
Active while we look at 2.0 port

**Comment by sepidehMS** [comment]
Fix was approved by shiproom and PR is merged in rel/2.0.0 (https://github.com/dotnet/corefx/pull/21117). Closing.
